### DISCUSSION TITLE
About the runtime module implementation

### DISCUSSION THREAD

**Discussion by cubxxw** [k3s-io/k3s#6478] [Q&A]
About the runtime module implementation
Hello, I want to make a `runtime` about k3s, and I want to realize a `rootfs`, which has good readability and can achieve basic functions. I want to package it, may I ask how can I achieve it?

🏄‍♂️ **about me** ：[sealer](https://github.com/sealerio/sealer)

**Comment by brandond** [comment]
I don't think I'm understanding your question. What specifically do you mean by making a runtime. Are you talking about packaging up a minimal Linux distribution to run K3s, similar to https://github.com/rancher/k3os? Or do you mean just the root filesystem, such as we provide with k3s via https://github.com/k3s-io/k3s-root ?

If you're trying to package k3s and all of its dependencies, there is no need to do so. Just grab one of our binary release artifacts; that has everything included in it. If you also need the images, grab one of the airgap tarballs.

**Reply by cubxxw** [reply]
I would like to create a runtime image, and then I would like to package all the k3s installation packages and dependencies. Maybe you have a point in that last statement, but your installation script is too long

**Reply by brandond** [reply]
There are no packages or dependencies. All you need is the binary. The install script just downloads it, and adds a systemd unit and other niceties. You might check out the documentation?

**Reply by cubxxw** [reply]
yep🏄‍♂️
As a sophomore student, your documentation is very difficult~
And your script is really long, ~ I think you can parse the following~
🫰🫰🫰

**Comment by cubxxw** [comment]
🧷 [http://sealer.cool/docs/getting-started/applications.html](http://sealer.cool/docs/getting-started/applications.html)

You can read this article. I want to build rootfs like this, but I don't understand where your configuration files or binaries or dependencies or other files are stored. @brandond

**Reply by brandond** [reply]
You can put the main k3s binary wherever you want. It will write content to /var/lib/rancher/k3s and /etc/rancher, as well as the normal locations used by containerd and kubelet for non-persistent files under /var/run.

**Comment by cubxxw** [comment]
Another question is about install step. AFAIK, k3s support join node by k3s agent k3s server?
Last, k3s support custom private registry, how to config this? and whether k3s could support external CRI?

**Reply by brandond** [reply]
> k3s support join node by k3s agent k3s server?

 Yes, `k3s agent` nodes can join the cluster of another node running `k3s server`. Is that what you're trying to ask?
 
 > k3s support custom private registry, how to config this? 
 
Have you had a chance to look at the documentation yet? https://docs.k3s.io/installation/private-registry
 
 > k3s could support external CRI?
 
 Yes, with --docker or --container-runtime-endpoint. The registries.yaml support mentioned above only works with embedded containerd though; if you bring your own CRI it is up to you to configure it properly.

**Comment by cubxxw** [comment]
@brandond 
I have read the official documentation and one question that puzzles me is that the new version of `k3s` is not prepared to use SQLite as the default metadata store. I want to ask next, if I choose k3s, so in this piece of how to choose, database ` curl - sfL https://get.k3s.io | sh - ` is the default etcd?
<br>

Another question about the fact that many people choose k3s because it is lightweight, why did you switch to etcd which is a bit more complicated as the default database?

**Reply by brandond** [reply]
> the new version of k3s is not prepared to use SQLite as the default metadata store.

That has not changed. sqlite is still the default. Where did you see something that made you think otherwise?

> why did you switch to etcd which is a bit more complicated as the default database

We haven't. Embedded etcd is only used if you want to have multiple servers without an external datastore, and requires starting k3s with `--cluster-init`. This is covered in the docs at https://docs.k3s.io/installation/ha-embedded.

**Comment by cubxxw** [comment]
@brandond 
Maybe you should take a look at this: https://docs.k3s.io/installation/ha-embedded

It tells me：
> Embedded etcd replaced experimental Dqlite in the K3s v1.19.1 release. This is a breaking change. Please note that upgrades from experimental Dqlite to embedded etcd are not supported. If you attempt an upgrade it will not succeed and data will be lost.

May I ask if you are going to use etcd by default? I would like to know your future plan.
> 
If you are not going to use etcd by default later, please tell me the default location of SQLite
>

**Reply by brandond** [reply]
dqlite is not sqlite. You can read more about it at https://github.com/canonical/dqlite. We used it only briefly for our embedded HA option, before replacing it with etcd. This statement says nothing about our dropping sqlite. We have no plans to make additional changes to the datastore options.

You should be able to find the db directory within the /var/lib/rancher/k3s path mentioned above.

**Comment by cubxxw** [comment]


**Reply by brandond** [reply]
I already responded to both of these questions at https://github.com/k3s-io/k3s/discussions/6478#discussioncomment-4159136

**Comment by cubxxw** [comment]
@brandond  Yes, sorry, I have read the documentation a lot lately and I have a rough idea of what the documentation means. I have tried every installation method so far and I have encountered a problem that I cannot find the answer to 🔽
 
'INSTALL_K3S_SYSTEMD_DIR' -- the directory where systemd services and environment files are installed, or use /etc/systemd/system as the default directory.

I want to change it and test it

```bash
curl -sfL https://get.k3s.io | INSTALL_K3S_MIRROR=cn \
  INSTALL_K3S_SYSTEMD_DIR=/opt/systemd \
  sh -
```

```bash
root@cubmaster01:/opt# pwd;cat k3s.service
/opt
[Unit]
Description=Lightweight Kubernetes
Documentation=https://k3s.io
Wants=network-online.target
After=network-online.target

[Install]
WantedBy=multi-user.target

[Service]
Type=notify
EnvironmentFile=-/etc/default/%N
EnvironmentFile=-/etc/sysconfig/%N
EnvironmentFile=-/opt//k3s.service.env
KillMode=process
Delegate=yes
# Having non-zero Limit*s causes performance problems due to accounting overhead
# in the kernel. We recommend using cgroups to do container-local accounting.
LimitNOFILE=1048576
LimitNPROC=infinity
LimitCORE=infinity
TasksMax=infinity
TimeoutStartSec=0
Restart=always
RestartSec=5s
ExecStartPre=/bin/sh -xc '! /usr/bin/systemctl is-enabled --quiet nm-cloud-setup.service'
ExecStartPre=-/sbin/modprobe br_netfilter
ExecStartPre=-/sbin/modprobe overlay
ExecStart=/usr/local/bin/k3s \
    server \
```
This is my first question. Yes, this file already exists, but I don't know the meaning of the content of this file. Do you have the explanation?
The second problem is that when I look up the service location of k3s, it is not the same as I expected. Look down:
![image](https://user-images.githubusercontent.com/86140903/204088913-d582cdd4-2fde-4567-995c-5f5f40a5bbfe.png)

I don't seem to understand the significance of this service location setting.

**Reply by brandond** [reply]
If you don't know what systemd unit files are for, you probably don't need to worry about it. Especially if you're running k3s in containers, where it is not expected that you would use a process supervisor like systemd.

**Comment by cubxxw** [comment]
@brandond 

Ok, I see. 
Whether k3s will allow me to extend features that I need in the future that are unique to kubernetes but that k3s does not have, or whether there is a way to smooth the transition from k3s to kubernetes, I believe this is also a consideration for many small and medium enterprises.
